File: backend/app/api/v1/wiki.py
"""知识库 Wiki 端点（v24）。

四件事：看目录与状态、读一页（带出处）、触发重建、清空。

**生成是异步的**（`TaskKind.WIKI` + 任务队列）：一次生成要跑"规划 + 每页一次
模型调用"，同步接口会占着连接几分钟。所以 POST 只入队并回任务 id，
前端轮询 `GET /knowledge-bases/{kb_id}/wiki` 的 `status`。

页面正文里的 `[n]` 与 `sources[].index` 一一对应——**出处是这一层的核心承诺**，
没有出处的自动 Wiki 没人敢信（调研报告 §3.5）。
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, status

from app.api.auth import WRITE, check_kb_scope, require_read, require_write
from app.api.v1.schemas import (
    WikiGenerateOut,
    WikiOverviewOut,
    WikiPageDetailOut,
    WikiPageOut,
    WikiSourceOut,
)
from app.core.services import Services, get_services
from app.services.api_key import Caller

router = APIRouter(tags=["wiki"])

# 页面/出处的组装放在这里，但**不 import 存储层的记录类型**（工程规范 §3.3 L1：
# 协议适配层只能转发 services/）。与 documents.py 的 `_to_out` 同一套写法——
# 参数不加类型标注，形状由 services 返回的对象保证。


def _page_out(record) -> WikiPageOut:  # type: ignore[no-untyped-def]
    return WikiPageOut(
        id=record.id,
        parent_id=record.parent_id,
        level=record.level,
        ord=record.ord,
        title=record.title,
        brief=record.brief,
        status=record.status,
        generated_at=record.generated_at,
    )


def _sources_out(services: Services, sources) -> list[WikiSourceOut]:  # type: ignore[no-untyped-def]
    """把出处的 ``document_id`` 解析成文档名（一次批量查，不逐条）。

    ``heading_path`` / ``page`` 不在这里回查 chunks——它们是生成时抄下来的快照，
    文档后来被删或重切也仍然指得准。
    """
    names = {
        record.id: record.name
        for record in services.documents.get_documents_by_ids(
            [item.document_id for item in sources]
        ).values()
    }
    return [
        WikiSourceOut(
            index=item.index,
            chunk_id=item.chunk_id,
            document_id=item.document_id,
            document_name=names.get(item.document_id, item.document_id),
            heading_path=item.heading_path,
            page=item.page,
        )
        for item in sources
    ]


@router.get(
    "/knowledge-bases/{kb_id}/wiki",
    response_model=WikiOverviewOut,
    summary="Wiki 目录与生成状态",
)
async def get_wiki(
    kb_id: str,
    services: Services = Depends(get_services),
    caller: Caller = Depends(require_read),
) -> WikiOverviewOut:
    check_kb_scope(services, caller, [kb_id])
    kb = services.knowledge_bases.get(kb_id)
    pages = services.wiki.pages(kb_id)
    state, last_error = services.wiki.status(kb_id)
    page_count, generated_at = services.wiki.stats(kb_id)
    return WikiOverviewOut(
        kb_id=kb_id,
        enabled=kb.wiki_enabled,
        status=state,
        page_count=page_count,
        generated_at=generated_at,
        model=None,
        last_error=last_error,
        pages=[_page_out(record) for record in pages],
    )


@router.get(
    "/wiki/pages/{page_id}",
    response_model=WikiPageDetailOut,
    summary="读一页 Wiki（正文 + 出处）",
)
async def get_wiki_page(
    page_id: str,
    services: Services = Depends(get_services),
    caller: Caller = Depends(require_read),
) -> WikiPageDetailOut:
    record = services.wiki.page(page_id)
    # 页面归属的知识库决定可见性：先取页面再判范围（同文档端点的顺序）
    check_kb_scope(services, caller, [record.kb_id])
    sources = services.wiki.sources(page_id)
    return WikiPageDetailOut(
        **_page_out(record).model_dump(),
        kb_id=record.kb_id,
        content_md=record.content_md,
        model=record.model,
        updated_at=record.updated_at,
        sources=_sources_out(services, sources),
    )


@router.post(
    "/knowledge-bases/{kb_id}/wiki/generate",
    response_model=WikiGenerateOut,
    status_code=status.HTTP_202_ACCEPTED,
    summary="重建这个知识库的 Wiki（异步）",
)
async def generate_wiki(
    kb_id: str,
    services: Services = Depends(get_services),
    caller: Caller = Depends(require_write),
) -> WikiGenerateOut:
    """入队一次重建。已在队列里时**返回同一个任务**（幂等，避免连点堆任务）。

    Wiki 形态没开的话由服务层回 409（`ConflictError`）并说明怎么开。
    """
    check_kb_scope(services, caller, [kb_id], need=WRITE)
    task = services.wiki.enqueue(kb_id)
    return WikiGenerateOut(task_id=task.id, kb_id=kb_id)


@router.delete(
    "/knowledge-bases/{kb_id}/wiki",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="清空这个知识库的 Wiki 页面",
)
async def clear_wiki(
    kb_id: str,
    services: Services = Depends(get_services),
    caller: Caller = Depends(require_write),
) -> None:
    """只删页面，**不动库形态开关**：用户可能只是想重来一次。"""
    check_kb_scope(services, caller, [kb_id], need=WRITE)
    services.wiki.clear(kb_id)

File: frontend/src/api/wiki.ts
/**
 * 知识库 Wiki 接口（对应后端 `/api/v1` 的 wiki 一组路由）。
 *
 * Wiki 是挂在某个知识库上的**一套百科式页面**：把库里已录入的内容整理成分层页面，
 * 每个要点都带原文出处。它有两种粒度，对应两组接口：
 * - **总览**（`getWiki`）：页面树 + 整体状态，页面只带标题/摘要，供左侧导航用；
 * - **单页**（`getWikiPage`）：某一页的完整正文与出处，按需拉取，避免一次传整本书。
 *
 * 生成是异步任务（后端返回 202 + `task_id`），所以界面拿到的状态里会有
 * `generating` 这一档，需要轮询总览直到它落定。字段与后端 schema 一一对应，
 * 改后端时同步改这里（与 `knowledgeBases.ts` 同一约定）。
 */

import { request } from './client'

/** 页面在导航树里的摘要。正文不在这里——树只负责"有哪些页、谁在谁下面"。 */
export interface WikiPageSummary {
  id: string
  /** 父页面 id；null = 顶层页面。`level` 与它是同一套层级信息的两种表达。 */
  parent_id: string | null
  /** 层级，从 0 开始（0 = 总览页，1 = 主题页）。后端已经算好，前端只按它缩进。 */
  level: number
  /** 同级排序序号，后端已按它排过序；前端再用它兜一次底。 */
  ord: number
  title: string
  /** 一句话摘要，列表与页头都用它。可能为空串。 */
  brief: string
  /**
   * 单页状态。整库生成时逐页落库，所以树里可能出现"有的页好了、有的页还在生成"。
   * 这是**页面级**状态，与总览的整库状态不是一回事。
   */
  status: 'ready' | 'generating' | 'failed'
  generated_at: string | null
}

/** 整库 Wiki 总览。 */
export interface WikiOverview {
  /** 这个库有没有开 Wiki（老库默认关）。关的时候整页应该给引导而不是空树。 */
  enabled: boolean
  /** 整库状态：idle 未生成 / generating 生成中 / ready 已生成 / failed 失败。 */
  status: 'idle' | 'generating' | 'ready' | 'failed'
  page_count: number
  generated_at: string | null
  /** 生成用的对话模型名。null = 还没生成过。 */
  model: string | null
  /** 最近一次失败原因。非空时界面要原样显示出来，而不是笼统说"失败了"。 */
  last_error: string | null
  pages: WikiPageSummary[]
}

/** 一条出处：正文里的 `[n]` 与它一一对应。 */
export interface WikiSource {
  /** 与正文里的 `[n]` 对应。 */
  index: number
  chunk_id: string
  document_id: string
  document_name: string
  /** 在原文里的标题路径（如「第三章 › 并发症」）。可能为空。 */
  heading_path: string | null
  /** 页码，PDF 才有；其它格式为 null。 */
  page: number | null
}

/** 单页详情。 */
export interface WikiPageDetail {
  id: string
  kb_id: string
  parent_id: string | null
  level: number
  title: string
  brief: string
  /** 正文 Markdown，含 `[n]` 引用标记（由 `renderAnswerWithCitations` 变成可点徽标）。 */
  content_md: string
  status: 'ready' | 'generating' | 'failed'
  model: string | null
  generated_at: string | null
  updated_at: string | null
  sources: WikiSource[]
}

/** 这个库的 Wiki 总览（页面树 + 状态）。 */
export function getWiki(kbId: string): Promise<WikiOverview> {
  return request(`/knowledge-bases/${kbId}/wiki`)
}

/** 单页正文与出处。 */
export function getWikiPage(pageId: string): Promise<WikiPageDetail> {
  return request(`/wiki/pages/${pageId}`)
}

/**
 * 触发整库生成/重新生成。
 *
 * 后端是**异步**的：立刻返回 202 与任务 id，真正的生成在后台跑。
 * 界面拿到 task_id 后不该傻等——重新拉一次总览拿到 `generating`，然后开始轮询。
 */
export function generateWiki(kbId: string): Promise<{ task_id: string }> {
  return request(`/knowledge-bases/${kbId}/wiki/generate`, { method: 'POST' })
}

/** 清空这个库已生成的 Wiki 页面（不删库、不动文档）。 */
export function clearWiki(kbId: string): Promise<void> {
  return request(`/knowledge-bases/${kbId}/wiki`, { method: 'DELETE' })
}
